File: starter/swift/Codegen/Assem.swift
//
//  Assem.swift
//  L1 Compiler
//  Assembly language
//  Author: Kaustuv Chaudhuri <kaustuv+@andrew.cmu.edu>
//  Modified By: Alex Vaynberg <alv@andrew.cmu.edu>
//  Modified: Frank Pfenning <fp@cs.cmu.edu>
//  Converted to OCaml by Michael Duggan <md5i@cs.cmu.edu>
//  Converted to Swift by Jonathan Burns <jaburns@andrew.cmu.edu>
//
//  Currently just a pseudo language with 3-operand
//  instructions and arbitrarily many temps
//
//  We write
//
//  BINOP  operand1 <- operand2,operand3
//  MOV    operand1 <- operand2
//

class Assem {

    typealias Program = [Instr]

    enum Reg: CustomStringConvertible {
        case eax
        
        var description: String {
            switch self {
            case .eax: return "%eax"
            }
        }
    }
    
    enum Operand: CustomStringConvertible {
        case imm(Int32)
        case reg(Reg)
        case temp(Temp)
        
        var description: String {
            switch self {
            case .imm(let val): return "$\(val)"
            case .reg(let r): return r.description
            case .temp(let t): return t.description
            }
        }
    }
    
    enum Operation: CustomStringConvertible {
        case add, sub, mul, div, mod

        var description: String {
            switch self {
            case .add: return "+"
            case .sub: return "-"
            case .mul: return "*"
            case .div: return "/"
            case .mod: return "%"
            }
        }

    }

    enum Instr: CustomStringConvertible {
        case binop(Operation, Operand, Operand, Operand)
        case mov(Operand, Operand)
        case directive(String)
        case comment(String)

        var description: String {
            switch self {
            case .binop(let oper, let d, let s1, let s2):
                return "\t\(d) <-- \(s1) \(oper) \(s2)\n"
            case .mov(let d, let s):
                return "\t\(d) <-- \(s)\n"
            case .directive(let str):
                return "\(str)\n"
            case .comment(let str):
                return "\t/*\(str)*/\n"
            }
        }
    }

}

File: Compiler/lab2/lib/languages/assem.ml
(* L1 Compiler
 * Assembly language
 * Author: Kaustuv Chaudhuri <kaustuv+@andrew.cmu.edu>
 * Modified By: Alex Vaynberg <alv@andrew.cmu.edu>
 * Modified: Frank Pfenning <fp@cs.cmu.edu>
 * Converted to OCaml by Michael Duggan <md5i@cs.cmu.edu>
 *
 * Currently just a pseudo language with 3-operand
 * instructions and arbitrarily many temps
 *)

open Core

type label = Symbol.t
type operand =
  | Imm of Int32.t
  | Reg of Reg.t
  | Temp of Temp.t

type compop = Less | Greater | Equal | Leq | Geq | Neq

type comparison  = 
    { lhs : operand
    ; op : compop
    ; rhs : operand
    }

type operation =
  | Add | Sub | Mul | Div | Mod
  | Bit_and | Bit_or | Bit_xor | Lshift | Rshift
  | Comp of compop

type instr =
  | Return
  | Binop of
      { op : operation
      ; dest : operand
      ; lhs : operand
      ; rhs : operand
      }
  | Mov of
      { dest : operand
      ; src : operand
      }
  | Branch of 
    { if_label : label
    ; else_label : label option
    ; after_label : label
    ; cond : [`Single of operand | `Comparison of comparison]
    }
  | Jump of label
  | Label of label
  | Directive of string
  | Comment of string


(* functions that format assembly output *)

let format_reg = Reg.format

let format_comp = function
| Less -> "<"
| Greater -> ">"
| Equal -> "=="
| Leq -> "<="
| Geq -> ">="
| Neq -> "!="
;;

let format_binop = function
| Add -> "+"
| Sub -> "-"
| Mul -> "*"
| Div -> "/"
| Mod -> "%"
| Lshift -> "<<"
| Rshift -> ">>"
| Bit_or -> "|"
| Bit_and -> "&"
| Bit_xor -> "^"
| Comp c -> format_comp c
;;

let format_operand = function
  | Imm n -> "$" ^ Int32.to_string n
  | Temp t -> Temp.name t
  | Reg r -> format_reg r Reg.Quad
;;

let format = function
  | Return -> "return"
  | Binop binop ->
    sprintf
      "%s <-- %s %s %s"
      (format_operand binop.dest)
      (format_operand binop.lhs)
      (format_binop binop.op)
      (format_operand binop.rhs)
  | Mov mv -> sprintf "%s <-- %s" (format_operand mv.dest) (format_operand mv.src)
  | Branch b -> sprintf "if %s then %s else %s."
    (match b.cond with
     | `Single op -> format_operand op
     | `Comparison c -> 
        sprintf "%s %s %s" 
        (format_operand c.lhs)
        (format_comp c.op)
        (format_operand c.rhs))
    (Symbol.name b.if_label)
    (match b.else_label with 
    | Some x -> Symbol.name x 
    | None -> Symbol.name b.after_label)
  | Jump l -> "goto " ^ Symbol.name l
  | Label l -> Symbol.name l
  | Directive dir -> dir
  | Comment comment -> sprintf "/* %s */" comment
;;
